src/components/fields/ArrayField.js

```diff
@@ -88,68 +94,157 @@ class ArrayField extends Component {
-                </div>
-                <div className="col-xs-2 array-item-remove text-right">
-                  <button type="button" className="btn btn-danger col-xs-12"
-                    tabIndex="-1"
-                    onClick={this.onDropIndexClick(index)}>Delete</button>
-                </div>
-              </div>
-            );
+            const itemIdSchema = toIdSchema(itemSchema, itemIdPrefix, definitions);
+            const itemUiSchema = (uiSchema.items instanceof Array) ? uiSchema.items[index] : uiSchema.items || {};
+            return this.renderArrayFieldItem({
+              index, itemSchema, itemIdSchema, itemErrorSchema,
+              removable: false,
+              itemData: items[index],
+              itemUiSchema
+            });
           })
+        }{
+          additionalItems ?
+            additionalItems.map((item, index) => {
+              index = index + itemSchemas.length;
+              const itemErrorSchema = errorSchema ? errorSchema[index] : undefined;
+              const itemIdPrefix = idSchema.id + "_" + index;
+              const itemIdSchema = toIdSchema(additionalSchema, itemIdPrefix, definitions);
+              return this.renderArrayFieldItem({
+                index, itemIdSchema, itemErrorSchema,
+                itemData: items[index],
+                itemSchema: additionalSchema,
+                itemUiSchema: uiSchema.additionalItems
+              });
+            }) : null
         }</div>
-        <div className="row">
-          <p className="col-xs-2 col-xs-offset-10 array-item-add text-right">
-            <button type="button" className="btn btn-info col-xs-12"
-              tabIndex="-1" onClick={this.onAddClick}>Add</button>
-          </p>
-        </div>
+        {
+          additionalSchema ? (
+            <AddButton onClick={this.onAddClick}/>
+          ) : null
+        }
       </fieldset>
     );
   }
+
+  renderArrayFieldItem({
+      index, removable=true, itemSchema, itemData,
+      itemUiSchema, itemIdSchema, itemErrorSchema
+  }) {
+    const {SchemaField} = this.props.registry.fields;
+    return (
+      <div key={index}>
+        <div className={removable ? "col-xs-10" : "col-xs-12"}>
+          <SchemaField
+              schema={itemSchema}
+              uiSchema={itemUiSchema}
+              formData={itemData}
+              errorSchema={itemErrorSchema}
+              idSchema={itemIdSchema}
+              required={this.isItemRequired(itemSchema)}
+              onChange={this.onChangeForIndex(index)}
+              registry={this.props.registry}/>
+        </div>
+        {
+          removable ?
+            <div className="col-xs-2 array-item-remove text-right">
+              <button type="button" className="btn btn-danger col-xs-12"
+                      tabIndex="-1"
+                      onClick={this.onDropIndexClick(index)}>Delete</button>
+            </div>
+          : null
+        }
+      </div>
```

Comment:
I'd really like to see these extracted and moved to their own stateless components.